Adding support for Lagrangian framework in the MMG utility

loumalouomega opened the pull request:
Basically that, in any case to work as supposed the interpolation of internal variables is required

loumalouomega committed:
Adding support for Lagrangian framework

loumalouomega committed:
Correcting tests to the new fluid interface

loumalouomega commented:
I corrected the tests, that weren't working after one push of the fluid application

RiccardoRossi commented on applications/MeshingApplication/custom_processes/metrics_levelset_process.h:
```diff
@@ -301,7 +301,7 @@ class ComputeLevelSetSolMetricProcess
-     * @param Interpolation: The equivalent enum
+     * @return Interpolation: The equivalent enum
      */
         
```
can u put a more descriptive name? reading the mehtod i don't understand what it is about (the doc is clear)

RiccardoRossi commented on applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -154,6 +155,7 @@ void AddCustomUtilitiesToPython()
     ("MmgUtility3D", init<ModelPart&>())
     .def(init<ModelPart&, const std::string>())
     .def(init<ModelPart&, const std::string, const unsigned int>())
+    .def(init<ModelPart&, const std::string, const unsigned int, const std::string>())
```
why don't you pass directly the enum in the constructor?
also wouldn't it start to make sense passing a "Parameters"?

loumalouomega replied:
I don't know how to pass an enum in Python, specially using parameters in json

RiccardoRossi commented on applications/MeshingApplication/custom_utilities/mmg_utility.h:
```diff
@@ -1050,6 +1058,11 @@ class MmgUtility
             }
             else
             {
+                if (mFramework == Lagrangian)
```
really not clear to me.

kratos nodes have X() and X0() as coordinates, related as
X() = X0() + node.GetSolutionStepValue(DISPLACEMENT_X)
what is being done here? you need to interpolate BOTH, X, X0 (and the DISPLACEMENT)

loumalouomega replied:
The displacement is already interpolated with the others variables. I assume that the coordinates are already computed, but I will interpolate them as well.

RiccardoRossi commented on applications/MeshingApplication/custom_utilities/mmg_utility.h:
```diff
@@ -2280,7 +2390,7 @@ class MmgUtility
         double* step_data = pNode->SolutionStepData().Data(step);
         
-        // NOTE: This just works with tetrahedron (you are going to have poblems with anything else)
+        // NOTE: This just works with tetrahedron (you are going to have problems with anything else)
```
why don't you make it generic? particularly since you would like to have prisms too

loumalouomega replied:
The problem with the prisms is that it is needed to change the the BinBasedFastPointLocator*, I will do that in a future, but it is an important change that requires some extra work. 

*The template of BinBasedFastPointLocator is BinBasedFastPointLocator<TDim>, and it just computes triangles and tetrahedron

RiccardoRossi replied:
ok, fine to me

RiccardoRossi commented on applications/MeshingApplication/custom_utilities/mmg_utility.h:
```diff
@@ -2244,6 +2292,68 @@ class MmgUtility
+        double& initial_coord_X = pNode->X0();
+        double& initial_coord_Y = pNode->Y0();
+        
+        const array_1d<double, 3>& node0_initial_coord = pElement->GetGeometry()[0].Coordinates() - pElement->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT, 0);
```
omit the 0 in the FastGetSolutionStepValue, it will be slightly faster

RiccardoRossi replied:
also a question here: if you assume, as you do, that nodes have displacement (which is something you should check at least for the first node), then you could use the interpolated value to reconstruct the original position.

i mean:
node.X0 = node.X - node.GetSolutionStepValue(DISPLACEMENT)

that would be faster, and consistent with the interpolation of displacements

loumalouomega replied:
Done

loumalouomega committed:
Appliying suggested modifications

Additionally correcting and changing the old KRATOS_THROW_ERROR with KRATOS_ERROR

loumalouomega commented:
I modified what @RiccardoRossi  suggested, now I am starting to implement the internal variables interpolation.

loumalouomega committed:
Important corrections

Nodes wasn't added directly to the submodelparts (solved), problems with no conditions produced errors in 2D (solved)

loumalouomega commented:
I did some corrections that were producing problems with submodelparts with just nodes and i corrected an error that occurred in 2D when the modelpart hasn't any condition

loumalouomega committed:
Forward to a internal variable interpolation while meshing

Forward to a internal variable interpolation while meshing

loumalouomega commented:
@RiccardoRossi I corrected some things that can be interested to add to the master branch, can you allow me to merge?, I will keep this branch until I figure out how to interpolate correctly internal variables

RiccardoRossi commented:
ok, to me you can push your changes

loumalouomega commented:
Thanks
